[Typechecker] Ban tuples with duplicate labels

theblixguy opened the pull request:
This PR bans tuples with duplicate labels. For example:

```swift
let f: (foo: Int, foo: Int) = (foo: 1, foo: 2)
```

Resolves [SR-8974](https://bugs.swift.org/browse/SR-8974)
Resolves rdar://problem/45218256

theblixguy committed:
[Typechecker] Ban tuples with duplicate labels

theblixguy commented:
cc @slavapestov @CodaFi

slavapestov reviewed (commented):
Nice!

What about tuple patterns?

```
if case (foo: let x, foo: let y) = zzz { ... }
```

Can you check enum declarations with duplicate labels in the payload type too?

theblixguy committed:
[Typechecker] Don't add empty identifiers to the set

theblixguy committed:
[Test] Adds a new test case

theblixguy commented:
You mean enum cases with duplicate labels? I added a new test case:

```swift
let zeroTuple = (0,0)
if case (foo: let x, foo: let y) = zeroTuple { print(x+y) }
```

can't get that to work with an enum (expression type 'T' is ambiguous without more context) though.

theblixguy committed:
[Test] Adds a new test case where we have an enum case with duplicate label

theblixguy committed:
Merge branch 'master' into fix/duplicate-tuple-labels

slavapestov commented on test/Constraints/tuple.swift:
```diff
@@ -289,3 +289,21 @@ func singleElementTuple() {
+if case (foo: let x, foo: let y) = zeroTuple { print(x+y) } // expected-error {{cannot create a tuple with a duplicate element label}} 
+// expected-warning@-1 {{'if' condition is always true}}
+
+enum BishBash { case bar(foo: Int, foo: String) }
```
Ideally we would diagnose the error here -- do we not diagnose duplicate function parameter names?

theblixguy replied:
@slavapestov Nope!

slavapestov replied:
Ok then. We do diagnose duplicate parameter names in functions, so maybe worth investigating this in a subsequent PR :)

slavapestov commented:
Did you want to look into patterns at some point too? It's a bit weird that this compiles:

```
q.swift:2:4: warning: 'if' condition is always true
if case (foo: let x, foo: let y) = (1, 2) { }
   ^
```

slavapestov commented:
@swift-ci Please smoke test

theblixguy commented:
@slavapestov Yeah, I’ll address that in a new PR where I’ll fix the enum case issue as well!

Seems like the current check needs some tweaking as the build is failing:

> /Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-test/branch-master/swift/stdlib/public/Darwin/ModelIO/ModelIO.swift:85:16: error: cannot create a tuple with a duplicate element label
20:47:52         __reset(with: array, count: array.count, atTimes: times, count: times.count)

Seems like an issue with ModelIO?

theblixguy committed:
[Typechecker] Do not check tuple labels for a function call

It's okay to have two function parameters with the same external label, as long as the internal labels are different. If not, we already emit a diagnostic for it, so no need to check

theblixguy commented:
Fixed. @slavapestov could you run the CI again? Thank you!

theblixguy committed:
[test] Update test

theblixguy commented:
Wait never mind, seems like simply checking for a function call won't always work. I need to get the parent of the tuple expression and check if it's a function call.

theblixguy committed:
[Typechecker] Check for CallExpr only, not ApplyExpr. Also, check for EnumElementDecl separately

theblixguy committed:
[Typechecker] Fix a crash

slavapestov commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -245,6 +250,42 @@ static void diagSyntacticUseRestrictions(TypeChecker &TC, const Expr *E,
+              break;
+            }
+
+            names.insert(name);
```
You can actually call insert() unconditionally and test the return value to see if the insertion suceeded instead

slavapestov commented:
@swift-ci Please smoke test

jrose-apple commented:
Who wants to bet someone is using these? :-(

@swift-ci Please test source compatibility

slavapestov commented:
Good point, we might have to turn this into a warning...

hamishknight commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -245,6 +250,42 @@ static void diagSyntacticUseRestrictions(TypeChecker &TC, const Expr *E,
+          }
+        }
+
+        if (!funcCallExpr || isEnumCase) {
```
What if this is a subscript index? I believe this will incorrectly diagnose `s[foo: 5, foo: 5]`. In addition, because `funcCallExpr` will be set when visiting other sub-expressions, I believe this won't catch things like:

```swift
func foo(_ fn: () -> Void) {}
foo { print((bar: "", bar: 5).bar) }
```

Ideally the check you want here is `if (!CallArgs.count(tupleExpr))` (which would let you merge this logic with the previous `if` block checking for single-element tuples). Although that will exclude enum case constructor calls, so you'd either want to check if the `Parent` is a `CallExpr` for an enum element decl, or ideally diagnose them separately on the decl.

theblixguy replied:
Oh yeah, I'll fix that once the source compatibility tests pass.

theblixguy replied:
Fixed 👍

hamishknight reviewed (commented):
Glad to see this getting fixed!

theblixguy committed:
[Typechecker] Use CallArgs for checking duplicate label & add new test cases

theblixguy commented:
Source compatibility passed, so it seems like no one is actually using it. I have addressed Hamish's comments - could someone invoke the CI again? Thank you!

jrose-apple commented:
Can you add to CHANGELOG, at least, then, since it could *theoretically* break someone?

theblixguy committed:
[Changelog] Adds an entry for SR-8974

theblixguy commented:
Done @jrose-apple 👍

jrose-apple commented on CHANGELOG.md:
```diff
@@ -26,6 +26,28 @@ CHANGELOG
+* [SR-8974][]:
+
+  Duplicate tuple element labels are no longer allowed, as they lead to
+  undefined behavior. For example:
```
Yikes! Is it really undefined? I would have expected it to always pick the first element in most cases.

theblixguy replied:
Yeah, that's what it does in most cases (not sure about its behaviour in past releases though). I should probably not call it "undefined" then - maybe just "because it leads to incorrect behaviour".

jrose-apple replied:
Sure, that sounds less scary.

theblixguy replied:
@jrose-apple Done 👍

jrose-apple commented on CHANGELOG.md:
```diff
@@ -26,6 +26,28 @@ CHANGELOG
+  will now be diagnosed as an error. 
+
+  Note: You can still use duplicate labels when declaring functions and
+  subscripts, as long as the internal labels are different. For example:
```
At some point this will be allowed for enum cases too…

theblixguy committed:
[Changelog] Tweak changelog entry

jrose-apple commented:
@swift-ci Please test

theblixguy commented:
Tests have passed 🎉 can I cherry pick this to 5.1 as well?

jrose-apple commented:
Seems reasonable, it's a pretty localized thing to check.
